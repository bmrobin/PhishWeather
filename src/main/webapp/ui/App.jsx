import React from 'react';
import Date from './date';
import { Show } from './show/show';
import { Weather } from './weather/weather';
import { Error } from './error/error';
import { getWeather } from '../services/weatherService';
import { lookupShowByDate } from '../services/phishAPIService';

export class App extends React.Component {
  
  constructor() {
    super();

    this.state = {
      date: {
        day: undefined,
        month: undefined,
        year: undefined
      },
      show: undefined,
      weather: undefined
    };

    this.dateChangeEvent = this.dateChangeEvent.bind(this);
    this.findShow = this.findShow.bind(this);
  }

  dateChangeEvent(dateProperty, value) {
    let newDate = this.state.date;
    switch (dateProperty) {
      case 'day':
        newDate.day = value;
        break;
      case 'month':
        newDate.month = value;
        break;
      case 'year':
        newDate.year = value;
        break;
      default:
        break;
    }
    this.setState({ date: newDate });
  }

  findShow() {
    let showResult, weatherResult;
    lookupShowByDate(this.state.date).then((data) => {
      showResult = data;
      this.setState({
        show: showResult,
        error: undefined
      });
      getWeather(this.state.date, this.state.show.location)
        .then((data) => {
          weatherResult = data;
          this.setState({
            weather: weatherResult
          });
        })
        .catch((error) => {
          let errorMessage;
          if (error.status === 0) {
            errorMessage = "An error occurred connecting with the backend service!";
          } else if (error.status === 404) {
            errorMessage = 'Weather for the Phish show was not found!';
          }
          this.setState({
            error: errorMessage
          });
        });
    });
  }

  render() {
    return (
      <div>
        <Date day={this.state.date.day}
              month={this.state.date.month}
              year={this.state.date.year}
              dateChange={this.dateChangeEvent} />
        <button id="get-date-id" type="button" onClick={this.findShow}>Lookup Date</button>
        {
          this.state.show &&
          <Show show={this.state.show} /> 
        }
        {
          this.state.weather &&
          <Weather weather={this.state.weather} />
        }
        {
          this.state.error &&
          <Error error={this.state.error} />
        }
      </div>
    );
  }
}
