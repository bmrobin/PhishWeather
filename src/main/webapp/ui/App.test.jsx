import React from 'react';
import { App } from './App';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

jest.mock('../date/Date.css');

describe("App component", () => {

  test("should initially only render date component", () => {
    let component = renderer.create( <App /> );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("should render show, weather, and error components in the app when present", () => {
    let show = {
      showdate: '2017-07-10',
      venue: 'Raleigh, NC'
    };
    let component = mount( <App /> );
    component.setState({show: show});
    expect(component.find('.show').exists()).toBe(true);

    let weather = {
      averageTempFahrenheit: 82,
      date: '2017-07-10'
    };
    component.setState({weather: weather});
    expect(component.find('#temperature-id').exists()).toBe(true);

    let error = 'error message';
    component.setState({error: error});
    expect(component.find('#error-id').exists()).toBe(true);
  });

});
