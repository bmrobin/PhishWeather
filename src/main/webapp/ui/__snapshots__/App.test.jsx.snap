// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should initially only render date component 1`] = `
<div>
  <div
    className="date"
  >
    <select
      id="month-id"
      onChange={[Function]}
      value={undefined}
    >
      <option
        default={true}
        value="null"
      >
        --Month--
      </option>
      <option
        value="01"
      >
        1
      </option>
      <option
        value="02"
      >
        2
      </option>
      <option
        value="03"
      >
        3
      </option>
      <option
        value="04"
      >
        4
      </option>
      <option
        value="05"
      >
        5
      </option>
      <option
        value="06"
      >
        6
      </option>
      <option
        value="07"
      >
        7
      </option>
      <option
        value="08"
      >
        8
      </option>
      <option
        value="09"
      >
        9
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="11"
      >
        11
      </option>
      <option
        value="12"
      >
        12
      </option>
    </select>
    <select
      id="day-id"
      onChange={[Function]}
      value={undefined}
    >
      <option
        value="null"
      >
        --Day--
      </option>
      <option
        value="01"
      >
        1
      </option>
      <option
        value="02"
      >
        2
      </option>
      <option
        value="03"
      >
        3
      </option>
      <option
        value="04"
      >
        4
      </option>
      <option
        value="05"
      >
        5
      </option>
      <option
        value="06"
      >
        6
      </option>
      <option
        value="07"
      >
        7
      </option>
      <option
        value="08"
      >
        8
      </option>
      <option
        value="09"
      >
        9
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="11"
      >
        11
      </option>
      <option
        value="12"
      >
        12
      </option>
      <option
        value="13"
      >
        13
      </option>
      <option
        value="14"
      >
        14
      </option>
      <option
        value="15"
      >
        15
      </option>
      <option
        value="16"
      >
        16
      </option>
      <option
        value="17"
      >
        17
      </option>
      <option
        value="18"
      >
        18
      </option>
      <option
        value="19"
      >
        19
      </option>
      <option
        value="20"
      >
        20
      </option>
      <option
        value="21"
      >
        21
      </option>
      <option
        value="22"
      >
        22
      </option>
      <option
        value="23"
      >
        23
      </option>
      <option
        value="24"
      >
        24
      </option>
      <option
        value="25"
      >
        25
      </option>
      <option
        value="26"
      >
        26
      </option>
      <option
        value="27"
      >
        27
      </option>
      <option
        value="28"
      >
        28
      </option>
      <option
        value="29"
      >
        29
      </option>
      <option
        value="30"
      >
        30
      </option>
      <option
        value="31"
      >
        31
      </option>
    </select>
    <select
      id="year-id"
      onChange={[Function]}
      value={undefined}
    >
      <option
        default={true}
        value="null"
      >
        --Year--
      </option>
      <option
        value="1988"
      >
        1988
      </option>
      <option
        value="1989"
      >
        1989
      </option>
      <option
        value="1990"
      >
        1990
      </option>
      <option
        value="1991"
      >
        1991
      </option>
      <option
        value="1992"
      >
        1992
      </option>
      <option
        value="1993"
      >
        1993
      </option>
      <option
        value="1994"
      >
        1994
      </option>
      <option
        value="1995"
      >
        1995
      </option>
      <option
        value="1996"
      >
        1996
      </option>
      <option
        value="1997"
      >
        1997
      </option>
      <option
        value="1998"
      >
        1998
      </option>
      <option
        value="1999"
      >
        1999
      </option>
      <option
        value="2000"
      >
        2000
      </option>
      <option
        value="2001"
      >
        2001
      </option>
      <option
        value="2002"
      >
        2002
      </option>
      <option
        value="2003"
      >
        2003
      </option>
      <option
        value="2004"
      >
        2004
      </option>
      <option
        value="2005"
      >
        2005
      </option>
      <option
        value="2006"
      >
        2006
      </option>
      <option
        value="2007"
      >
        2007
      </option>
      <option
        value="2008"
      >
        2008
      </option>
      <option
        value="2009"
      >
        2009
      </option>
      <option
        value="2010"
      >
        2010
      </option>
      <option
        value="2011"
      >
        2011
      </option>
      <option
        value="2012"
      >
        2012
      </option>
      <option
        value="2013"
      >
        2013
      </option>
      <option
        value="2014"
      >
        2014
      </option>
      <option
        value="2015"
      >
        2015
      </option>
      <option
        value="2016"
      >
        2016
      </option>
      <option
        value="2017"
      >
        2017
      </option>
    </select>
  </div>
  <button
    id="get-date-id"
    onClick={[Function]}
    type="button"
  >
    Lookup Date
  </button>
</div>
`;
